File: Week 6 - Python/sentimental-credit/credit.py
from cs50 import get_string
import re


number = get_string("Number: ")
n = int(number)
s1, s2, S, r, a = 0, 0, 0, 0, 0

while n > 0:
    r = n % 10
    s1 += r
    n = (n - r) // 10
    r = n % 10
    a = 2 * r

    while a > 0:
        b = a % 10
        s2 += b
        a = (a - b) // 10
        b = a % 10
        s2 += b
        a = a // 10

    S = s1 + s2
    n = n // 10

if S % 10 == 0:
    if len(number) == 15 and (re.match("34", number) or re.match("37", number)):
        print("AMEX")
    elif len(number) == 16 and (re.match("51", number) or re.match("52", number) or re.match("53", number) or re.match("54", number) or re.match("55", number)):
        print("MASTERCARD")
    elif (len(number) == 13 or len(number) == 16) and re.match("4", number):
        print("VISA")
    else:
        print("INVALID")
else:
    print("INVALID")

File: Week 1 - C/credit/credit.c
#include <cs50.h>
#include <stdio.h>

int card_length(long);
long starting_digits(long);
int main(void)
{
    long N = get_long("Number: ");
    int s1 = 0, s2 = 0, S = 0, r = 0, a = 0;
    long n = N;
    while (n > 0)
    {
        r = n % 10;
        s1 += r;
        n = (n - r) / 10;
        r = n % 10;
        a = 2 * r;
        while (a > 0)
        {
            int b = 0;
            b = a % 10;
            s2 += b;
            a = (a - b) / 10;
            b = a % 10;
            s2 += b;
            a = a / 10;
        }
        S = s1 + s2;
        n = n / 10;
    }

    if (S % 10 == 0)
    {
        if (card_length(N) == 15 && (starting_digits(N) == 34 || starting_digits(N) == 37))
        {
            printf("AMEX\n");
        }
        else if (card_length(N) == 16 && (starting_digits(N) == 51 || starting_digits(N) == 52 || starting_digits(N) == 53 ||
                                          starting_digits(N) == 54 || starting_digits(N) == 55))
        {
            printf("MASTERCARD\n");
        }
        else if ((card_length(N) == 13 || card_length(N) == 16) && (starting_digits(N) / 10 == 4))
        {
            printf("VISA\n");
        }
        else
        {
            printf("INVALID\n");
        }
    }
    else
    {
        printf("INVALID\n");
    }
    return 0;
}

int card_length(long n)
{
    int i = 0;
    while (n > 0)
    {
        n = n / 10;
        i++;
    }
    return i;
}

long starting_digits(long n)
{
    int length = card_length(n);
    for (int i = 0; i < length - 2; i++)
    {
        n = n / 10;
    }
    return n;
}
